### DISCUSSION TITLE
Move to ‘clap’ for parsing CLI arguments

### DISCUSSION THREAD

**Discussion by udsamani** [dathere/qsv#463] [General]
We are currently using 'docopt' for parsing arguments. It has following issues in my opinion

- Our files get filled with USAGE strings. It would be really nice to have it broken and let the library itself take care of.
- docopt is no longer maintained.
- docopt library does not handle parsing subarguments and we have to manually worry about it. For example if you take a look at 'apply.rs' we parse sub arguments for 'operation' command. Ideally library should take care of it.

**Comment by jqnatividad** [comment]
@udsamani I thought about this myself (https://github.com/jqnatividad/qsv/issues/61), and I'm open to merging a PR to do so.

qsv actually uses a [fork of docopt with some clippy/perf tweaks](https://github.com/docopt/docopt.rs/compare/master...jqnatividad:docopt.rs:perf-clippy-2021), and for qsv's purposes - seems "good enough."

I particularly like how the USAGE strings work, and how it's automagically parsed by docopt. But yeah - its unmaintained. 

Perhaps, we can leverage the same techniques cargo applied to migrate to clap in 2018 (https://github.com/rust-lang/cargo/pull/5129)?

Anyways, I'll reopen the ticket and put it in the backlog and would welcome PRs.

**Comment by jqnatividad** [comment]
Now that clap v4.0 has been released, its time to revisit this issue.

https://github.com/clap-rs/clap/blob/master/CHANGELOG.md#400---2022-09-28

**Comment by jqnatividad** [comment]
Still sticking with `docopt` for now as it allows us to have usage text that not only acts as a way to specify CLI parameters, it also enables:
- verbose, readable, properly formatted usage text, replete with examples, that's easy to maintain
- keep documentation with the code, with one contiguous usage string
- users can just read the usage text on GitHub with their browser, they don't need to download and install qsv first to invoke it with the --help option, to read its usage text.
- have the Available Commands table on the README.md link directly to each command's source code

Granted, `clap` is more modern, is actively maintained, and is the de facto CLI parsing standard, but migrating qsv to it won't be trivial and we'll have to break up the usage string to setup clap, so it'll be interspersed with code.

**Comment by jqnatividad** [comment]
qsv now uses an optimized fork of docopt.rs - [qsv_docoopt](https://github.com/jqnatividad/docopt.rs#qsv_docopt).

As qsv has fully embraced docopt's self-documenting nature that neither clap nor structopt can provide, we'll maintain qsv_docopt going forward not only for general upkeep but to ensure it takes advantage of Rust's latest innovations & features.

Besides, the two main known issues of docopt.rs (https://github.com/docopt/docopt.rs/issues/207 and https://github.com/docopt/docopt.rs/issues/152) are not critical to how we use it in qsv.

For 207, its not relevant as qsv doesn't use docopt in that way. 152 is still valid though, and we've added it to qsv_docopt's backlog - https://github.com/jqnatividad/docopt.rs/issues/2
